### PR TITLE
feat(storage): use standard PUT replacement semantics

### PR DESCRIPTION
**Pull request by Fermionic-Lyu**
## What

Storage uploads now follow S3-style PUT semantics:

- `PUT /api/storage/buckets/:bucket/objects/*` creates or replaces the object at the exact key and returns **200**.
- Metadata writes use one `INSERT ... ON CONFLICT (bucket, key) DO UPDATE` in the caller's RLS context. Replacements preserve `uploaded_by`.
- Presigned upload strategies preflight the same create-or-replace write in a rolled-back savepoint before issuing a URL; `confirm-upload` performs the same atomic metadata write.
- `autoKey` remains available for callers that explicitly want a server-generated unique key. The POST auto-key upload route is unchanged.
- Backend and dashboard auto-renaming are removed. No `upsert` query/body parameter is introduced.

## Why

The previous backend behavior silently renamed a conflicting key (`photo.png` → `photo (1).png`). That differs from S3 `PutObject`, where a successful PUT to an existing key replaces the current object. Callers that want a new unique object should opt into the auto-key flow instead.

## Permissions

End-user writes still run under `storage.objects` RLS. A new key must satisfy the INSERT policy; a conflicting key also takes the UPDATE conflict arm and must satisfy the applicable UPDATE/visibility policies. A denied replacement fails before provider bytes are written. Project-admin/API-key behavior is unchanged by this PR.

## Breaking change

Uploading the same explicit key twice now replaces the first object instead of auto-renaming it or returning 409. Use the POST auto-key route or an SDK `uploadAuto` flow when a new unique key is required.

The paired TypeScript SDK change in InsForge/InsForge-sdk-js#114 should be narrowed to `autoKey`/`uploadAuto`; its proposed `upsert` option is no longer part of this API contract.

## Testing

- Format, lint, typecheck, and production build pass on Node 22.
- Backend: **151 files passed / 1809 tests passed** (3 files and 17 tests skipped by configuration).
- Dashboard: **35 files / 132 tests passed**.
- PR CI passes: pre-merge, migrations, lint/format, unit, integration, dashboard/UI, and repository E2E.
- Multi-arch Docker image built successfully from `210ffd877`: https://github.com/InsForge/InsForge/actions/runs/29784564348
- Updated deterministic fixture: InsForge/agent-e2e#12.
- The fixture's Storage and S3 modules fully pass against that image, including non-owner replacement denial and owner same-key replacement: https://github.com/InsForge/agent-e2e/actions/runs/29784710756. The overall run is red only because this PR branch predates the unrelated `/api/ai/config` endpoint now asserted by current agent-e2e main.

> [!NOTE]
> ## What
>
> Storage uploads now follow S3-style PUT semantics:
>
> - `PUT /api/storage/buckets/:bucket/objects/*` creates or replaces the object at the exact key and returns **200**.
> - Metadata writes use one `INSERT ... ON CONFLICT (bucket, key) DO UPDATE` in the caller's RLS context. Replacements preserve `uploaded_by`.
> - Presigned upload strategies preflight the same create-or-replace write in a rolled-back savepoint before issuing a URL; `confirm-upload` performs the same atomic metadata write.
> - `autoKey` remains available for callers that explicitly want a server-generated unique key. The POST auto-key upload route is unchanged.
> - Backend and dashboard auto-renaming are removed. No `upsert` query/body parameter is introduced.
>
> ## Why
>
> The previous backend behavior silently renamed a conflicting key (`photo.png` → `photo (1).png`). That differs from S3 `PutObject`, where a successful PUT to an existing key replaces the current object. Callers that want a new unique object should opt into the auto-key flow instead.
>
> ## Permissions
>
> End-user writes still run under `storage.objects` RLS. A new key must satisfy the INSERT policy; a conflicting key also takes the UPDATE conflict arm and must satisfy the applicable UPDATE/visibility policies. A denied replacement fails before provider bytes are written. Project-admin/API-key behavior is unchanged by this PR.
>
> ## Breaking change
>
> Uploading the same explicit key twice now replaces the first object instead of auto-renaming it or returning 409. Use the POST auto-key route or an SDK `uploadAuto` flow when a new unique key is required.
>
> The paired TypeScript SDK change in InsForge/InsForge-sdk-js#114 should be narrowed to `autoKey`/`uploadAuto`; its proposed `upsert` option is no longer part of this API contract.
>
> ## Testing
>
> - Format, lint, typecheck, and production build pass on Node 22.
> - Backend: **151 files passed / 1809 tests passed** (3 files and 17 tests skipped by configuration).
> - Dashboard: **35 files / 132 tests passed**.
> - PR CI passes: pre-merge, migrations, lint/format, unit, integration, dashboard/UI, and repository E2E.
> - Multi-arch Docker image built successfully from `210ffd877`: https://github.com/InsForge/InsForge/actions/runs/29784564348
> - Updated deterministic fixture: InsForge/agent-e2e#12.
> - The fixture's Storage and S3 modules fully pass against that image, including non-owner replacement denial and owner same-key replacement: https://github.com/InsForge/agent-e2e/actions/runs/29784710756. The overall run is red only because this PR branch predates the unrelated `/api/ai/config` endpoint now asserted by current agent-e2e main.
> 
> #### Changes since #1760 opened
>
> - Removed server-side automatic key generation from upload strategy flow and implemented client-side file renaming logic using incremental suffix pattern [69e51a1]
> - Changed HTTP status code for `confirm-upload` endpoint from 201 to 200 to match direct PUT route semantics [69e51a1]
> - Updated documentation and OpenAPI specification across all language variants to remove `autoKey` field references and redirect users to POST endpoint for server-generated unique keys [69e51a1]
> - Added `listingCache` to batch upload logic in dashboard `BucketsPage.uploadFiles` helper to cache `storageService.listObjects` results per base filename [badbb88]
> - Updated documentation across English, Spanish, Traditional Chinese, and Simplified Chinese variants to specify that uploading to an existing object key replaces the current object in place [badbb88]
> - Added explanatory comments to migration `036_storage-third-party-auth-support.sql` documenting PUT create-or-replace semantics with INSERT ... ON CONFLICT [badbb88]
> 
>

### CODE REVIEW

**Review by jwfing — COMMENTED** [review]
## Summary

Solid, well-documented change that replaces API-layer auto-rename with standard PUT semantics (409 on conflict, `?upsert=true` to replace) and correctly relocates the friendly rename UX to the dashboard client — the RLS UPDATE-policy gate is the right mechanism and is applied consistently across the three write paths. No blocking issues found.

## Requirements context

No `/docs/superpowers/` or `docs/specs/` spec/plan exists in this repo (the `docs/` tree is product/marketing MDX only), so this was assessed against the PR description, the linked breaking-change note, and the existing storage RLS contract in `backend/src/infra/database/migrations/036_storage-third-party-auth-support.sql`. The implementation matches the stated intent.

## Findings

### Critical
(none)

### Suggestion

**Software engineering / functionality — `putObject` upsert relies on an unchecked error path where the other two paths defensively check `rowCount`** — `backend/src/services/storage/storage.service.ts:195-226`
The `confirmUpload` replace (`:824-836`) and the `upload-strategy` RLS probe (`:646-664`) both run a plain `UPDATE ... WHERE bucket/key` and treat `rowCount === 0` as "RLS hid the row → 403". But the `putObject` upsert path uses `INSERT ... ON CONFLICT (bucket, key) DO UPDATE` and does **not** check anything — its entire cross-user-overwrite guarantee rests on PostgreSQL raising `42501` when the existing row fails the UPDATE `USING (uploaded_by = jwt.sub)` policy (then mapped to 403 by `mapObjectWriteError`). This is in fact the correct/documented PG behavior for `ON CONFLICT DO UPDATE` (the arbiter-index-located row can't be silently filtered, so a USING failure errors rather than no-ops), and `tests/local/test-storage-rls.sh:216-234` covers it end-to-end (`Bob PUT upsert → 403`, `Alice's blob untouched`). The unit tests, however, mock the DB and can only assert the happy-path `ON CONFLICT` SQL shape — none exercises the `42501 → 403` mapping *for the putObject path specifically*. Given that a regression here (e.g. a future policy edit, or PG treating it as a no-op) would silently overwrite another user's blob via the subsequent `provider.putObject`, a short defensive comment on `insertObject` pinning *why* no `rowCount` check is needed — plus a unit test feeding a `{ code: '42501' }` rejection through `putObject({ upsert: true })` — would harden the most security-sensitive path. Low blast radius today because the shell test guards it, hence Suggestion not Critical.

**Functionality — dashboard auto-rename silently caps at 1000 keys / 3 retries** — `packages/dashboard/src/features/storage/pages/BucketsPage.tsx:189-213`
`nextAvailableObjectKey` is fed a single `listObjects(..., { limit: 1000 }, baseName)` page. In a bucket with >1000 objects sharing the base name, the highest existing `name (N).ext` counter can fall outside the page, so the computed candidate collides, 409s, and after 3 attempts the raw conflict error is surfaced to the user instead of a rename. Acceptable degradation for an edge case, but worth either logging the give-up or noting the cap.

### Information

- **`req.query.upsert === 'true'` is exact-string only** — `backend/src/api/routes/storage/index.routes.ts:315`. `?upsert=1`, `?upsert=TRUE`, or a bare `?upsert` are treated as `false`. Safe default and consistent with the body-param `=== true` checks, just worth documenting for SDK authors.
- **409-on-existing discloses key existence across the RLS boundary** — `storage.service.ts:172-179`. Because `objectKeyExists` runs on the admin pool, a caller learns a key is taken even when RLS hides the owning row. This is inherent to a globally-unique keyspace with per-user RLS and is unchanged from the old behavior (auto-rename leaked the same fact), and it matches S3/Supabase semantics — noting only for completeness, not a regression.
- **Route error-handling asymmetry** — the PUT (`index.routes.ts:349-351`) and `upload-strategy` (`:751-753`) catches added an explicit `if (error instanceof AppError) next(error)` guard, but `confirm-upload` (`:799-811`) did not. It's still correct — the 403/409 `AppError`s from `confirmUpload` carry their own `statusCode` and none of the string-match branches (`not found` / `already confirmed` / `exceeds the configured maximum`) match their messages — so this is purely a consistency nit.
- **Import cleanup verified** — `escapeRegexPattern` removal is clean (no remaining references), `escapeSqlLikePattern` is still used by the list/search paths (`:467,475,1173`), and both `putObject` callers were updated to the new `{ object, replaced }` shape. Good scope discipline; the diff does only what the PR describes.

### Positive notes
- Ownership preservation on replace (`uploaded_by` left out of every `SET` clause, asserted by unit tests) correctly mirrors `upsertS3Object` (`:967-985`) and prevents an admin/API-key replace from NULLing a user's ownership.
- INSERT-before-provider-write ordering (and the etag persistence moved *outside* the tx at `:248-270`) is preserved for the non-upsert path, so a provider failure still can't leave an orphan row.
- Good test coverage overall: 6 new unit tests for the 409/probe/upsert branches, a rewritten RLS shell test for the new contract, and the new `nextAvailableObjectKey` helper is well-tested including regex-escaping and dotfile edge cases.

## Verdict

**approved** (informational — zero Critical findings; the Suggestion/Information items are non-blocking). Posting as a COMMENT per the bot's non-authority-to-approve rule; explicit GitHub approval remains a human action.

**Review by jwfing — DISMISSED** [review]
LGTM - approved.

File: packages/dashboard/src/features/storage/pages/BucketsPage.tsx [hunk 1]
```diff
@@ -171,6 +173,44 @@ export default function BucketsPage() {
...
+        );
+        try {
+          return await uploadObject({ bucket, objectKey: candidateKey, file });
+        } catch (error) {
+          if (!isConflict(error)) {
+            throw error;
+          }
+          // Another upload took the candidate between listing and PUT; re-list.
+          lastError = error;
+        }
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/storage/pages/BucketsPage.tsx]: Fixed in d11717ef6 — retries now record every 409-rejected candidate and feed them back into the next-key computation, so each attempt is guaranteed to advance even when the capped listing misses the blocker. I stopped short of full pagination: exhausting pages of a substring match in a large bucket costs unbounded round-trips for an edge case, and the candidate-tracking already guarantees forward progress within the retry budget.

File: packages/dashboard/src/features/storage/pages/BucketsPage.tsx [hunk 2]
```diff
@@ -171,6 +173,44 @@ export default function BucketsPage() {
...
+      const baseName = lastDotIndex > 0 ? file.name.slice(0, lastDotIndex) : file.name;
+      let lastError: unknown;
+      for (let attempt = 0; attempt < 3; attempt++) {
+        const { objects } = await storageService.listObjects(bucket, { limit: 1000 }, baseName);
+        // Force the conflicting name into the list: the server said it exists,
+        // even if the (paginated) search listing missed it.
+        const candidateKey = nextAvailableObjectKey(
+          [...objects.map((object) => object.key), file.name],
+          file.name
+        );
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/storage/pages/BucketsPage.tsx]: Right — the substring search can crowd out the counter variants. Fixed in d11717ef6 by tracking rejected candidates across retries instead of trusting the listing to be complete; each retry now provably advances.

File: packages/dashboard/src/features/storage/pages/BucketsPage.tsx [hunk 3]
```diff
@@ -171,6 +173,44 @@ export default function BucketsPage() {
...
+      }
+
+      const lastDotIndex = file.name.lastIndexOf('.');
+      const baseName = lastDotIndex > 0 ? file.name.slice(0, lastDotIndex) : file.name;
+      let lastError: unknown;
+      for (let attempt = 0; attempt < 3; attempt++) {
+        const { objects } = await storageService.listObjects(bucket, { limit: 1000 }, baseName);
+        // Force the conflicting name into the list: the server said it exists,
+        // even if the (paginated) search listing missed it.
+        const candidateKey = nextAvailableObjectKey(
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/storage/pages/BucketsPage.tsx]: Fixed in d11717ef6 — conflicted candidates are preserved across attempts and included in the next computation.

File: backend/src/services/storage/storage.service.ts [hunk 4]
```diff
@@ -148,80 +148,40 @@ export class StorageService {
...
     originalKey: string,
     file: Express.Multer.File,
-    hasApiKey: boolean = false
-  ): Promise<StorageFileSchema> {
+    hasApiKey: boolean = false,
+    options: { upsert?: boolean } = {}
+  ): Promise<{ object: StorageFileSchema; replaced: boolean }> {
     this.validateBucketName(bucket);
     this.validateKey(originalKey);
+    const upsert = options.upsert === true;
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/storage/storage.service.ts]: Fixed in d11717ef6 — direct strategies now return the uploadUrl with upsert=true appended when upsert was requested, so redeeming the URL verbatim works. Presigned URLs intentionally carry no flag: S3 PUT replaces natively and the gate runs at strategy issuance plus confirm-upload.

File: backend/src/services/storage/storage.service.ts [hunk 5]
```diff
@@ -743,44 +797,65 @@ export class StorageService {
...
-         VALUES ($1, $2, $3, $4, $5, $6, 'rest', $7)`,
-        [bucket, key, fileSize, metadata.contentType || null, finalEtag, userId, uploadedAt]
-      );
-    if (hasApiKey || ctx?.role === 'project_admin') {
-      await runWithRootAccess(this.getPool(), insertObjectRow);
-    } else {
-      if (!ctx) {
-        throw new AppError('Forbidden', 403, ERROR_CODES.STORAGE_PERMISSION_DENIED);
+    const writeObjectRow = async (db: PoolClient) => {
+      if (rowExists) {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/storage/storage.service.ts]: Fixed in d11717ef6 — the upsert confirm path is now a single INSERT ... ON CONFLICT DO UPDATE (ownership excluded from the SET clause), same shape as putObject, so concurrent upsert confirms can no longer race into a 409.

File: backend/src/api/routes/storage/index.routes.ts [hunk 6]
```diff
@@ -705,7 +712,7 @@ router.post(
   async (req: AuthRequest, res: Response, next: NextFunction) => {
     try {
       const { bucketName } = req.params;
-      const { filename, contentType, size } = req.body;
+      const { filename, contentType, size, upsert, autoKey } = req.body;
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/storage/index.routes.ts]: Fixed in d11717ef6 — openapi/storage.yaml now documents the upsert query param on PUT with 200/403/409 responses, the upsert/autoKey fields on upload-strategy, and the upsert field on confirm-upload.

File: backend/src/services/storage/storage.service.ts [hunk 7]
```diff
@@ -606,14 +638,34 @@ export class StorageService {
...
+          );
+          if ((probe.rowCount ?? 0) === 0) {
+            // RLS hides the row from this caller's UPDATE policy — they may
+            // not replace someone else's object.
+            throw new AppError(
+              `You do not have permission to replace "${key}" in bucket "${bucket}"`,
+              403,
+              ERROR_CODES.STORAGE_PERMISSION_DENIED
+            );
+          }
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/storage/storage.service.ts]: Real but self-healing: the window is between the admin-pool existence check and the RLS probe, and a client retry lands on the create path and succeeds. Disambiguating would need a re-check loop inside the probe for a race that resolves itself on retry, so leaving as is.

File: backend/src/services/storage/storage.service.ts [hunk 8]
```diff
@@ -148,80 +148,40 @@ export class StorageService {
...
-    const finalKey = await this.generateNextAvailableKey(bucket, originalKey, this.getPool());
+    const finalKey = originalKey;
+    const exists = await this.objectKeyExists(bucket, finalKey);
+    if (exists && !upsert) {
+      throw new AppError(
+        `Object "${finalKey}" already exists in bucket "${bucket}". Pass upsert=true to replace it.`,
+        409,
+        ERROR_CODES.STORAGE_ALREADY_EXISTS
+      );
+    }
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/storage/storage.service.ts]: Acknowledged, but not fixing: the only way to derive replaced from the write itself is RETURNING (xmax = 0), and under RLS, RETURNING applies the table's SELECT policies to the returned row — that re-couples write success to read visibility, which this file deliberately avoids (see the uploaded_at comment in putObject; it broke end-user uploads before). The cost of the race is a mislabeled 201 in a hairline window with no content-correctness impact, since last-writer-wins is the documented upsert contract.

File: backend/src/services/storage/storage.service.ts [hunk 9]
```diff
@@ -283,15 +270,43 @@ export class StorageService {
...
-      url: this.buildObjectUrl(bucket, finalKey, etag || uploadedAt),
+      object: {
+        bucket,
+        key: finalKey,
+        size: file.size,
+        mimeType: file.mimetype,
+        uploadedAt: uploadedAt.toISOString(),
+        url: this.buildObjectUrl(bucket, finalKey, etag || uploadedAt),
+      },
+      replaced: exists,
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/storage/storage.service.ts]: Same root cause and trade-off as the sibling comment on the exists flag: deriving replaced from the upsert result requires RETURNING, which re-applies SELECT RLS to writes — a coupling this codebase deliberately avoids. Content is correct either way; only the 200/201 label can be stale in the race window.

File: packages/dashboard/src/features/storage/helpers.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  const escapeRegex = (value: string) => value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+  const counterRegex = new RegExp(
+    `^${escapeRegex(baseName)} \\((\\d+)\\)${escapeRegex(extension)}$`
+  );
+
+  let highestCounter = 0;
+  for (const key of existingKeys) {
+    const match = key.match(counterRegex);
+    if (match) {
+      highestCounter = Math.max(highestCounter, parseInt(match[1], 10));
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/storage/helpers.ts:29]: Not fixing: this needs a real object whose suffix exceeds 2^53 — e.g. photo (9007199254740993).png — already in the bucket, and even then the failure mode is a visible 409 toast after retries, not data loss or overwrite. bigint handling here isn't warranted for that input.

**Comment by Fermionic-Lyu** [comment]
Pushed d11717ef6 addressing the review round:

- The putObject upsert path now documents exactly why no rowCount check is needed (the ON CONFLICT arm raises 42501 when the existing row fails the UPDATE policy's USING — it cannot silently no-op), and a new unit test drives a 42501 rejection through `putObject({ upsert: true })` and asserts 403 with no provider blob write. (jwfing's suggestion.)
- upload-strategy and confirm-upload bodies are validated with the shared zod schemas. One deliberate behavior nuance: `size: 0` is now accepted on confirm-upload where `if (!size)` used to reject it — the provider HEAD size check is authoritative anyway. (CodeRabbit's nitpick.)
- Direct upload-strategy URLs now carry `?upsert=true` when requested, the upsert confirm is a single race-free `ON CONFLICT` write, dashboard rename retries track rejected candidates, and `openapi/storage.yaml` documents the new contract — details on the inline threads.

Declined with reasons on the inline threads: deriving `replaced` from the write via RETURNING (re-couples writes to SELECT RLS), the deleted-row-vs-403 disambiguation in the strategy probe (self-healing race), and bigint suffix counters (requires a 2^53 suffix to exist).

Deterministic Fixture E2E is being re-run against a rebuilt `v2.2.7-storage-upsert` image; will post the run link when it completes.

**Comment by Fermionic-Lyu** [comment]
Deterministic Fixture E2E re-ran against the rebuilt `v2.2.7-storage-upsert` image (commit d11717ef6) and passed, including the new 409/upsert fixture assertions: https://github.com/InsForge/agent-e2e/actions/runs/29769707488

**Review by jwfing — COMMENTED** [review]
## Review: standard PUT/upsert semantics; move auto-rename to dashboard

**Summary:** A careful, well-tested change that replaces API-layer auto-rename with standard PUT semantics (409 on conflict, `?upsert=true` to replace), gating end-user replaces through the existing `storage.objects` RLS UPDATE policies and preserving ownership — no Critical issues found.

### Requirements context
No matching spec/plan found — `InsForge/InsForge` has no `docs/superpowers/` or `docs/specs/` directory (only product docs under `docs/`). Assessed against the PR description, the linked SDK PR (#114), and the surrounding storage code. Intent (PUT idempotency + RLS-gated replace + client-side rename) is clearly stated and the diff matches it.

---

### Critical
(none)

### Suggestion

**Functionality — presigned `confirm-upload` always returns `201`, even on an upsert replace** — `backend/src/api/routes/storage/index.routes.ts:805`
The PR establishes the contract "`200` on replace, `201` on create". The direct `PUT` route honors this (`index.routes.ts:350` via the `replaced` flag). But the presigned flow's `confirm-upload` hard-codes `successResponse(res, fileInfo, 201)` and `confirmUpload` returns a `StorageFileSchema` with no `replaced` field (`storage.service.ts:895-902`), so a client on an S3/cloud backend that replaces via presigned+confirm gets `201` and no way to tell it was a replacement. Consider threading a `replaced` boolean through `confirmUpload` and returning `200` on the conflict arm, to keep both upload paths consistent. Non-blocking — the PR only formally promises the 200/201 split for `PUT /objects/*`.

**Software engineering — no unit test for the end-user (RLS) `confirmUpload` upsert 403 path** — `backend/tests/unit/storage-object-is-visible.test.ts:612`
`putObject` has an excellent test pinning the `42501 → 403` conflict-arm behavior for an authenticated non-owner (line 462). `confirmUpload` upsert is only tested for the `project_admin` root path (line 612). The end-user `withUserContext` upsert path (the one the RLS gate actually protects) shares `mapObjectWriteError` with `putObject`, so it's low-risk, but a mirrored test would lock in that the presigned replace is genuinely RLS-gated the same way the direct PUT is.

### Information

- **Performance — dashboard rename retry issues up to 3 sequential list queries per conflicting file** — `packages/dashboard/src/features/storage/pages/BucketsPage.tsx:198-215`. For bulk uploads where many files collide this is `N × up-to-3` list calls, all sequential. Fine for interactive dashboard use and the loop is correctly bounded (feeds rejected candidates back in to guarantee forward progress), just noting the cost.
- **Security — no new exposure.** New user input (`upsert`, `autoKey`, object key) is schema-validated (`@insforge/shared-schemas`) and parameterized; the key still runs through `validateKey` traversal checks; the RLS savepoint probe correctly proves write permission *before* issuing a presigned URL (which bypasses RLS at redeem); no secrets/PII newly logged or returned; no auth checks weakened. The `?upsert=true` flag cannot be abused to overwrite another user's object — both the strategy-time probe and the confirm-time `ON CONFLICT` run under `withUserContext`, so a non-owner hits `42501 → 403` either way (verified by the RLS shell test rewrite).
- **Compatibility — signatures are backward compatible.** `putObject`/`getUploadStrategy`/`confirmUpload` gained trailing optional `options` params with defaults; the only other `getUploadStrategy` caller (`deployment.service.ts:235`) calls the *provider* method, not the service, so it's unaffected. `generateNextAvailableKey` is fully removed with no dangling references. The breaking change (old SDKs relying on rename now get 409) is explicitly documented with release sequencing against #114.
- **Note:** the "Provider write stays inside the transaction — the tx rolls back, no orphan row remains" comment (`storage.service.ts:222-225`) is accurate for the end-user `withUserContext` path but not the admin/API-key `runWithRootAccess` path, which autocommits per statement. This is pre-existing (unchanged from base) and out of scope for this PR — flagging only for awareness.

---

### Verdict: `approved`
No Critical findings. Standard PUT semantics are implemented correctly, ownership-preservation and RLS gating are sound and pinned by both unit tests and the rewritten `test-storage-rls.sh`, and the OpenAPI docs accurately reflect the new 200/201/403/409 contract. The two Suggestions are non-blocking polish. (Informational only — a human still gives the explicit GitHub approval.)

**Review by jwfing — DISMISSED** [review]
LGTM - approved.

**Comment by Fermionic-Lyu** [comment]
Pushed 69e51a1c3 addressing this round. Since the branch moved to standard S3 PUT semantics (create-or-replace on the exact key, no `?upsert`), most of the bot comments target the removed `?upsert` design and are stale — the direct-URL upsert flag, the 200/201 `replaced` races, "concurrent upsert confirm 409", the staged-upload `_uploads/` cleanup (that staging commit was reverted), and the "plain Error at :841" pre-check are all gone with the code they referenced. The docs comments suggesting a "default 409 / pass upsert=true" also predate the rework; the docs correctly describe replace-always.

What actually changed in 69e51a1c3:

- **Dashboard auto-rename restored.** The S3-semantics rework had dropped it, which left the dashboard silently overwriting on a duplicate filename. Since PUT no longer 409s, the dashboard now proactively lists and picks the next free `name (N).ext` before uploading (reserving keys within a batch). Client-only; the API never renames.
- **`autoKey` removed** from shared-schemas, the route, and `getUploadStrategy`. It wasn't an S3 concept and duplicated the existing `POST /objects` server-key path. The SDK's `uploadAuto` now mints the key client-side and uses the standard upload path (InsForge/InsForge-sdk-js#114). `generateObjectKey` stays for `POST /objects`.
- **confirm-upload now returns 200** (was 201) to match the direct PUT — both finalize a create-or-replace, and neither reads back to distinguish create vs replace.

OpenAPI and all REST/TS doc locales updated to match. Local checks green (backend tsc, storage service/route + dashboard helper + storage-feature suites, shared-schemas + dashboard typecheck, lint). Holding the E2E re-run — the fixture exercises PUT directly and isn't affected by the dashboard/docs/autoKey changes; happy to dispatch it if you'd prefer.

**Review by jwfing — COMMENTED** [review]
## Code Review — feat(storage): standard PUT replacement semantics

**Summary:** Replaces the backend's silent auto-rename with S3-style create-or-replace PUT semantics via a single `INSERT ... ON CONFLICT (bucket, key) DO UPDATE` executed in the caller's RLS context, moves the friendly auto-rename to the dashboard client, and updates docs/OpenAPI/schemas to match — a clean, well-tested change with no blocking issues.

### Requirements context
No matching spec/plan found — `docs/superpowers/` and `docs/specs/` do not exist in this repo (consistent with the InsForge main repo layout). Assessed against the PR description, the paired SDK PR (#114), and the surrounding storage code.

---

### Critical
(none)

### Suggestion

**Performance / efficiency (software engineering)** — `packages/dashboard/src/features/storage/pages/BucketsPage.tsx:188-234`
`uploadWithAutoRename` issues one `listObjects(..., { limit: 1000 }, baseName)` per file inside the sequential upload loop. For a multi-file batch that's N list round-trips. Since `takenInBatch` already reserves in-batch keys, and the search is a substring match, you could hoist a single listing per batch (or per distinct baseName) and reuse it. Non-blocking — this is an admin dashboard flow and correctness is unaffected.

**Functionality** — `packages/dashboard/src/features/storage/pages/BucketsPage.tsx:181-189`
The client-side rename reads up to 1000 matching keys. In the (unlikely) case a bucket has >1000 keys sharing a base name, a higher-numbered counter could be missed and the chosen key could collide — which now silently *replaces* rather than 409s. The TOCTOU race between list and PUT is already called out in the code comment; the 1000-key ceiling is the same class of accepted trade-off but isn't noted. Consider a comment or a defensive check. Low blast radius.

### Information

**Functionality / RLS nuance** — `backend/src/services/storage/storage.service.ts:166-184`, `690-749`
`INSERT ... ON CONFLICT DO UPDATE` enforces the INSERT policy's `WITH CHECK` against the proposed row *and* the UPDATE policy's `USING`/`WITH CHECK` on conflict. For the migration's default owner policies this is exactly right (create gated by INSERT, replace gated by owner UPDATE, `uploaded_by` preserved). Projects that ship an *asymmetric* custom policy (e.g. a more restrictive INSERT than UPDATE) may find an owner's replacement gated by the INSERT `WITH CHECK` too. Worth a sentence in the storage RLS docs alongside the existing `upsertS3Object` note (`storage.service.ts:857-869`), since the same class of caveat applies.

**Docs** — `openapi/storage.yaml` (confirm-upload 404 example)
The 404 example emits `error: "UPLOAD_NOT_FOUND"`, but the route maps the `not found` message to `ERROR_CODES.STORAGE_NOT_FOUND` (`backend/src/api/routes/storage/index.routes.ts:791-793`). Appears pre-existing; flagging only for accuracy.

---

### What's solid
- **Software engineering:** Strong test coverage for the new behavior — backend unit tests cover in-place replacement with ownership preserved, RLS-denied replacement mapped to 403 *with the provider blob untouched*, the rolled-back upload-strategy probe, and confirm-upload replace/deny paths (`backend/tests/unit/storage-object-is-visible.test.ts`). The shell RLS test now asserts hidden-key replacement rejection + owner same-key replace (`test-storage-rls.sh`), e2e status codes updated 201→200, and the new dashboard helper has its own unit suite (`helpers.test.ts`). Removed `generateNextAvailableKey` has no dangling references; `escapeRegexPattern` import cleaned up.
- **Functionality:** PUT now returns 200 for both create and replace (matches S3 `PutObject`); `confirmUpload` mirrors PUT and the stale "already confirmed" 409 path is removed consistently across service, route, and OpenAPI; POST auto-key route (`generateObjectKey`) is genuinely unchanged; the etag persist stays outside the user tx to avoid orphaning blobs on a failed UPDATE. The paired SDK narrowing (#114) is correctly called out in the description.
- **Security:** All writes remain parameterized and run under `withUserContext` RLS for end users; a denied create/replace raises 42501 → 403 *before* any provider bytes are written; the presigned-upload permission probe is always rolled back in a savepoint; `escapeSqlLikePattern` guards LIKE queries; new request bodies are now validated with zod (`uploadStrategyRequestSchema`, `confirmUploadRequestSchema`) — a net tightening. No new secrets logged, no auth checks weakened.
- **Performance:** Net reduction in DB work — the old dedup `SELECT` + regex scan is replaced by one atomic upsert. No new N+1 or blocking work on the backend hot path (the only added round-trips are client-side, noted above).

### Verdict
`approved` — no Critical findings. Suggestions and Information items are non-blocking; the author may address or defer them. (This is the bot's assessment; explicit GitHub approval remains a separate human action.)

**Review by jwfing — DISMISSED** [review]
LGTM - approved.

**Comment by Fermionic-Lyu** [comment]
Pushed badbb884e addressing the latest round:

- **Docs contradiction** (cubic): the upload-strategy section no longer routes REST users to `POST /objects` for auto-key uploads — that endpoint's own section is marked "(Deprecated)" as part of this doc's consistent "use the strategy flow" stance, so cross-referencing it was the inconsistency. The line now just states that a PUT to an existing key replaces in place. All four REST locales.
- **Dashboard 1000-key ceiling** (cubic + jwfing): documented as an accepted trade-off alongside the existing TOCTOU note — with replace-always a missed higher counter replaces rather than 409s, but it needs >1000 keys sharing a base name in one bucket. Left as a comment rather than paginating, per the low blast radius.
- **N listings per batch** (jwfing): `uploadWithAutoRename` now caches one listing per base name per batch instead of re-listing per file.
- **RLS asymmetric-policy nuance** (jwfing): added to the migration 036 policy comment — ON CONFLICT checks both the INSERT WITH CHECK (create) and the UPDATE policy (replace), so a custom policy stricter on INSERT than UPDATE can also gate an owner's replacement.

Local checks green (backend tsc + storage/migration unit tests, dashboard tsc/lint/storage suites). No behavior change to the API or the RLS gate — docs, a dashboard optimization, and comments.

**Review by jwfing — APPROVED** [review]
LGTM - approved.
